### ISSUE TITLE
[Feature] Display manually added versions of modifications in the modification version list.

### ISSUE THREAD

**Issue by Artemonim** [KSP-CKAN/CKAN#3082]
Problem
-------
1. When adding a modification to the GameData folder manually, you can't use CKAN to find out which version you are using.
   1. Does CKAN see the version and can it offer to install the update if it finds it later?
2. I tried to install a [modification](https://github.com/c1usta/KS3P/releases/tag/6.1.2-b3) via "File" -> "Import downloaded mods...", but CKAN just doesn't do anything if the modification is already in its list.

Suggestions
-----------
1. If CKAN sees the mod installed not through it and can recognize its version, display the version in the list of versions.
2. When trying to import a version of a mod that exists in the CKAN catalog...
   1. If it is possible to recognize the version being added and it already exists in the catalog, suggest that the user install the version from the catalog.
   2. If a non-existent version was added to GameData, then add It to the list of mod versions.
      1. If CKAN can't recognize the version, ask the user to specify it manually. This is necessary both for the user's convenience and for the ability to update the modification via CKAN later.

![2020-06-17_10-02-03](https://user-images.githubusercontent.com/36363466/84865966-c1634680-b081-11ea-93c0-fcb905f144a4.png)

**Comment by DasSkelett** [comment]
> When adding a modification to the GameData folder manually, you can't use CKAN to find out which version you are using.
Does CKAN see the version [...]?

No it doesn't. The most we could do is try parsing the mod version from the version file, if present. We could look at implementing that, however it wouldn't work reliably, and the version numbers would likely differ in format from the version numbers from other releases known to CKAN.

> [...]  and can it offer to install the update if it finds it later

This is basically #949 (and many many duplicates). We can't do exactly that, however with https://github.com/KSP-CKAN/CKAN/pull/3043 you will be able to re-install the mod in CKAN without having to remove the files from GameData first.

---

Now to the import function. I think there's a bit of a misconception of what the import functionality actually does (not only you, also generally. We should try to make it clearer in the GUI):
The import functionality is there to be able to install mods (and mod versions) CKAN _knows about_ (that are indexed) without re-downloading them if you have, for some reason, already downloaded them manually.
It _can not_ install mods (and mod versions) that are not already indexed.

It detects the mod (version) you want to install by calculating the file hash and searching for matching ones in the registry. If it finds one or more matches, it imports the downloaded file into the cache and then installs it according to the already known install instructions for this mods version.

> I tried to install a modification via "File" -> "Import downloaded mods...", but CKAN just doesn't do anything if the modification is already in its list.

It probably didn't do nothing, instead it reported `Not found in index: KS3P-*.zip` in the status bar at the bottom. Something we should improve to give it more visibility.
And as explained above, it didn't do it because the mod release is _in_ the registry, but because it _isn't_.

> If it is possible to recognize the version being added and it already exists in the catalog, suggest that the user install the version from the catalog.

See above, you can _only_ import mod versions known to the registry.

> If a non-existent version was added to GameData, then add It to the list of mod versions.

See above

> If CKAN can't recognize the version, ask the user to specify it manually.
This is necessary both for the user's convenience and for the ability to update the modification via CKAN later.

To get the update functionality working (which is already not possible due to the reasons above), the user would have to input the **exact** version as it would be known to CKAN, including any transformations CKAN does, like adding epochs or adding leading `v`s if the mod has a flaky history, so that CKAN can compare the versions and find out if there are newer versions available. There is basically a zero chance that users would know what to do and that they would get it right.

**Comment by Artemonim** [comment]
What if the import opens a menu where you can choose which mod the version belongs to?
Or add a button like "import your version" to the list of modification versions.

**Comment by DasSkelett** [comment]
> What if the import opens a menu where you can choose which mod the version belongs to?
> Or add a button like "import your version" to the list of modification versions.

This would still leave the problem of how CKAN should get the version number of this mod release.
.version files are unreliable and most likely not formatted the same way as the version numbers we get from GitHub/SpaceDock. User input is even less reliable, users don't know how they should format the version.
And a problem I completely overlooked in my previous comment: how should CKAN know how to install this unindexed version? There are no install instructions available for it.
Similarly, all the other metadata are missing, most importantly the KSP compatibility and relationships.

In short, trying to implement a feature to install unindexed mods and mod releases would open a massive can of worms.

**Comment by HebaruSan** [comment]
Yeah, this isn't do-able due to reasons given by @DasSkelett above. CKAN needs to know many things about a mod that are not contained in the ZIP file. That's the reason NetKAN exists in the first place.
